typed-racket-lib/typed-racket/typecheck/def-binding.rkt

```diff
@@ -107,8 +107,9 @@
              (mk-ignored-quad e))))
 
      (define sname-is-constructor? (and (or extra-constr-name (free-identifier=? sname constructor-name)) #t))
-     (define constr (or extra-constr-name constr^))
+     (define constr (or extra-constr-name constructor-name))
```

Comment:
@samth the other meaningful change is here. The constructor name from the repackaged struct info by `request/typed` are not actually exported as the constructor, so we use the actual constructor name received by `dtsi*`.

Reply:
It is very likely that we can simplify the code for require/typed for structs, but we should do that in this PR